nbuild/stdenv/fetch.py

```diff
@@ -34,32 +35,48 @@ def fetch_url(
 
     if not _check_file(path, md5, sha1, sha256):
         ilog(f"Fetching {url}")
-        req = requests.get(url, stream=True)
-        with open(path, 'wb') as file:
-            for chunk in req.iter_content(chunk_size=4096):
-                file.write(chunk)
+        if url.startswith("http"):
+            download_direct(url, path)
+        elif url.startswith("ftp"):
+            download_ftp(url, path)
+        else:
+            flog(f"Unknown protocol to download file from url {url}")
+            exit(1)
         clog(f"Fetch done. Stored at {path}")
         if not _check_file(path, md5, sha1, sha256):
             flog(
                 "Downloaded file's signature is invalid. "
                 "Please verify the signature(s) in the build manifest "
                 "and the authenticity of the given link."
             )
-            print(f"Url: {url}\nmd5: {md5}")
             exit(1)
     else:
         clog(f"Using cache at {path}")
 
 
+def download_direct(url, path):
+        req = requests.get(url, stream=True)
+        with open(path, 'wb') as file:
+            for chunk in req.iter_content(chunk_size=4096):
+                file.write(chunk)
+
+
+def download_ftp(url, path):
+    with urllib.request.urlopen(url) as response, \
+         open(path, 'wb') as out_file:
+        data = response.read()
```

Comment:
Can' you do fixed-sized read instead? (Like in `download_direct()`)
Because your code requires to place the whole file in memory before writing it, which is very memory-consuming.